### PR TITLE
Change MultiNodeEvaluator to object patching from proxy pattern

### PR DESCRIPTION
**Pull request by kuenishi**
With Chainer #3952 current proxy implementation cannot handle
class attributes of Chainer's latest Evaluator.name, thus move
to just wrap the evaluate() method with multi-node support.

With this and #149 merged, ChainerMN will work with Chainer 4.0.

Note: consequence of this issue is, validation evaluator cannot have it's name `validation` and key `validation/main/loss` and `validation/main/accuracy` does not exist; which has two unexpected effects:

1. instead the evaluator overwrites `main/loss` and `main/accuracy`, printing no result in LogReport
2. due to the unexpected lack of `validation/main/loss` or accuracy user codes may fail (and following functionality such as saving actual models, might fail)

### CODE REVIEW

**Comment by kuenishi** [comment]
Although CircleCI is failing, but it is on test phase and not ready to use in reviewing and testing branches. This pull request is ready for review.

File: chainermn/extensions/multi_node_evaluator.py [hunk 1]
```diff
@@ -1,40 +1,36 @@
-def create_multi_node_evaluator(actual_evaluator, communicator):
+import types
+
+
+def create_multi_node_evaluator(evaluator, communicator):
```
Comment (iwiwi) [inline chainermn/extensions/multi_node_evaluator.py]: I agree that `evaluator` might be a better argument name as the name `actual_evaluator` somewhat discloses the implementation of `create_multi_node_evaluator`, which is not good. On the other hand, we want to follow Chainer's API compatibility rule (http://chainer.readthedocs.io/en/stable/compatibility.html) as much as possible. I think this change conflicts the rule. What do you think?

File: chainermn/extensions/multi_node_evaluator.py [hunk 2]
```diff
@@ -1,40 +1,36 @@
...
+        local_mean_dict = self._mn_original_evaluate()
+        global_mean_dict = {
+            name:
+            self._mn_communicator.mpi_comm.allreduce(
+                value) / self._mn_communicator.size
+            for name, value in sorted(local_mean_dict.items())
+        }
+        return global_mean_dict
+
+    evaluator.evaluate = types.MethodType(new_evaluate, evaluator)
```
Comment (iwiwi) [inline chainermn/extensions/multi_node_evaluator.py]: Seemingly this code works for both Python 2 and 3 now, but I'm a bit conscious of using `types.MethodType`, which is not documented in the official Python 3 document (https://docs.python.org/3/library/types.html#types.MethodType). How about using `six.create_bound_method` (https://pythonhosted.org/six/#six.create_bound_method)

**Comment by kuenishi** [comment]
I agree this changes behaviour of `create_multi_node_evaluator` which is not suitable for its name and the new behaviour can even be considered not compatible. However, I still think patching the evaluator object is the best way to fix Chainer 4.0.0 compatibility issue and I propose following actions on this:

- Add new patching function with new name like `bless_evaluator_for_multi_node` or sth like that.
- For `create_multi_node_evaluator`, we have several possible options:
  - Option 1: We would say for 1.1 it does not work with Chainer 4.0.0b2 or later (works fine with b1), But mark the function as to-be-deprecated at 1.1, leaving the bug as it is. And deprecate it at 2.0.
  - Option 2: We add a kludgy patch to it as a quick fix but mark as to-be-deprecated at 1.1 and at 2.0.
  - Option 3: apply a kludgy patch that fix this issue and not add evaluator-patching new feature.

A kludgy patch I mentioned above would be something like this:
```python
    class MultiNodeEvaluator(type(actual_evaluator)):
  ...
        def __setattr__(self, attr_name, value):
            if hasattr(self.actual_evaluator.__class__, attr_name):
                setattr(self.actual_evaluator.__class__, attr_name, value)
            else:   
                setattr(self.actual_evaluator, attr_name, value)
```

File: chainermn/extensions/multi_node_evaluator.py (hunk 2 above)
Comment (kuenishi) [reply chainermn/extensions/multi_node_evaluator.py]: I agree that and will replace with six in the next commit, after　we concluded on compatibility discussion.

**Comment by kuenishi** [comment]
After discussion with @iwiwi patching approach is fine, but just preserved the name `actual_evaluator` . Also, fixed `types.MethodType` but to use six. Will be included in 1.1 release with the new design.

File: chainermn/extensions/multi_node_evaluator.py [hunk 3]
```diff
@@ -1,40 +1,37 @@
...
+
     Args:
-        actual_evaluator: evaluator
+        actual_evaluator: evaluator to be patched
             (e.g., ``chainer.training.extensions.Evaluator``)
         communicator: ChainerMN communicator
 
     Returns:
-        The multi node evaluator based on ``actual_evaluator``.
+        The multi node evaluator based on ``evaluator``.
```
Comment (iwiwi) [inline chainermn/extensions/multi_node_evaluator.py]: Could you also fix this line?

**Comment by iwiwi** [comment]
Thanks, I will merge after Travis tests

**Pull request merged**